storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -814,7 +814,12 @@ private static int numUsedWorkers(SchedulerAssignment assignment) {
 
     private boolean auditAssignmentChanges(Map<String, Assignment> existingAssignments,
                                            Map<String, Assignment> newAssignments) {
-        assert existingAssignments != null && newAssignments != null;
+        if (existingAssignments == null) {
```

Comment:
Any reason why we don't use `Preconditions` here? It seems more elegant to me.

Reply:
assert/Precondition seems more concise - except for the unchecked exception being thrown.
In this specific instance (which is also throwing unchecked exception), maybe it is better to treat null map as an empty map.

Reply:
Isn't this `Preconditions.checkNotNull(existingAssignments);` what we wanted?

Reply:
Looked at this code more specifically. The asserts/preconditions seem superfluous. 
This method is private. first check is too late, i.e. already used multiple times before being passed, and the second param is instantiated before the call.

Reply:
What about other places? Can we use Preconditions everywhere?

Reply:
Throwing unchecked exceptions is generally not considered good practice. So using Preconditions would be not be considered good practice. Unchecked Exceptions should be thrown only in rare cases. In this case, it is a private method, totally in control of the class scope, and therefore no reason to throw unchecked exceptions. 

With regard to checking for null - that will not be considered defensive programming. Defense would be to check for null and then do something about it - i.e. set to to an empty list instead. The idea being that the method can handle parameters can can be "bad" in some sense. In this specific private method, that defense is not required.

Throwing unchecked exception is not a defense. 
last few lines here: https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

Reply:
Definitely need critical review of code that is throwing unchecked exception (including AssertionError()) and look for options  - or complete removal - even in the changed code. 

At best, "assert" usage should be viewed as lazy coding in private methods. And should not be used in protected or public methods.
https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html (note the language that it should not be used in public methods).

This does not mean that private methods cannot throw Exceptions. They can and very often do. But those are checked exceptions (i.e. present in method signature).

The goal of this change is to remove asserts. Replace with proper checking where possible, throw checked Exception where appropriate. Reason is quite straightforward - assert statements may not be executed and gives a false impression that something is being done.

Reply:
Sorry I am a little confused. So 

What's the difference between 
`assert existingAssignments != null && newAssignments != null;`
and 
`assert (TopologyStatus.ACTIVE == initStatus || TopologyStatus.INACTIVE == initStatus);`

Why the first one should be deleted and the second should be changed?

Reply:
First one is not required as it is never true (i.e a superfluous check in a private method - and is never activate at runtime - since it was in an assert to begin with.

Second one is a legitimate check that should be done at runtime. But was in an assert earlier and not activate earlier, but should be active.

Reply:
assertion is enabled at integration tests 
https://github.com/apache/storm/blob/master/integration-test/config/storm.yaml#L39-L41

Reply:
Correct, but only in test. This is fairly standard use of assert in test. But putting this in main code is not. 

If it a legitimate error condition then it should be checked. If the test needs to check something, then it should be in the test class.
Otherwise, it is gives the wrong impression that the asseret is checking something at runtime, when it typically isn't.

That is the premise of this change.